util.ts

```diff
@@ -1,18 +1,126 @@
+import { DataWriteOptions, FileManager, TFile, TFolder, Vault, normalizePath } from "obsidian";
+import { PickedFile } from "filesystem";
+
+
 function escapeRegex(str: string): string {
 	return str.replace(/[.?*+^$[\]\\(){}|-]/g, '\\$&');
 }
 
-const ILLEGAL_CHARACTERS = '\\/:*?<>\"|';
-const ILLEGAL_FILENAME_RE = new RegExp('[' + escapeRegex(ILLEGAL_CHARACTERS) + ']', 'g');
+const ILLEGAL_FILENAME_CHARACTERS = '\\/:*?<>\"|';
+const ILLEGAL_FILENAME_RE = new RegExp('[' + escapeRegex(ILLEGAL_FILENAME_CHARACTERS) + ']', 'g');
+
+const ILLEGAL_TAG_CHARACTERS = '\\:*?<>\"|!@#$%^&()+=\`\'~;,.';
+const ILLEGAL_TAG_RE = new RegExp('[' + escapeRegex(ILLEGAL_TAG_CHARACTERS) + ']', 'g');
+
+// Finds any non-whitespace sections starting with #
+const POTENTIAL_TAGS_RE = new RegExp(/(#[^ ^#]*)/, 'g');
 
 export function sanitizeFileName(name: string) {
 	return name.replace(ILLEGAL_FILENAME_RE, '');
 }
 
+/**
+ * Searches a string for characters unsupported by Obsidian in the hashtag body and returns a snaitized string.
+ * If the # symbol is included at the start or anywhere else it will be removed.
+ */
+export function sanitizeHashtag(name: string): string {
+	let tagName = name.replace(ILLEGAL_TAG_RE, '');
+	tagName = tagName.split(' ').join('-');
+	if(!isNaN(tagName[0] as any)) {
+		tagName = '_' + tagName;
+	}
+	return tagName;
+}
+
+/**
+ * Searches a string for hashtags that include characters unsupported in hashtags by Obsidian.
+ * Returns a string with those hastags normalised.
+ */
+export function sanitizeHashtags(str: string): string {
+	const newStr = str.replace(POTENTIAL_TAGS_RE, (str: string) : string => {
+		return '#' + sanitizeHashtag(str);
+	});
+	return newStr;
+}
+
 export function genUid(length: number): string {
 	let array: string[] = [];
 	for (let i = 0; i < length; i++) {
 		array.push((Math.random() * 16 | 0).toString(16));
 	}
 	return array.join('');
 }
+
+/**
+ * Retrieves a reference to a specific folder in a vault. Creates it first if it doesn't exist.
+ * Must pass in the relevant vault.
+ */
+export async function getOrCreateFolder(folderPath: string, vault: Vault): Promise<TFolder> {
```

Comment:
I've made a helper function to cover this use case in FormatImporter that should help! (latest master)